File: scripts/validate-github-workflows.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'find'
require 'open3'
require 'tmpdir'
require 'yaml'

ROOT = File.expand_path('..', __dir__)
GITHUB_DIR = File.join(ROOT, '.github')
FORBIDDEN_ACTION_REFS = %w[master main latest].freeze
FORBIDDEN_RUNTIME_FLAGS = %w[
  FORCE_JAVASCRIPT_ACTIONS_TO_NODE20
  FORCE_JAVASCRIPT_ACTIONS_TO_NODE22
  FORCE_JAVASCRIPT_ACTIONS_TO_NODE24
].freeze

# Production workflows must not pass by using placeholder credentials, dummy data,
# warning-only gates, skipped validations, or swallowed command failures.
FORBIDDEN_TEXT_PATTERNS = {
  /ci_dummy|dummy_|fake_|not-configured|your-project\.supabase\.co|your-anon-key|your-public-anon-key|your-publishable-key/i => 'placeholder or dummy credential/data detected',
  /continue-on-error:\s*true/i => 'continue-on-error is forbidden for production workflow gates',
  /skip_.*=true|skipped.*missing secrets|warning::.*skipped/i => 'skip-on-missing-secrets pattern detected',
  /\|\|\s*true/ => 'command failure swallowed with || true'
}.freeze

errors = []
workflow_files = Dir.glob(File.join(GITHUB_DIR, '**', '*.{yml,yaml}')).sort

if workflow_files.empty?
  errors << 'No GitHub YAML files found under .github/.'
end

def relative(path)
  path.delete_prefix(ROOT + '/')
end

def each_node(node, path = [], &block)
  yield(node, path)
  case node
  when Hash
    node.each { |key, value| each_node(value, path + [key], &block) }
  when Array
    node.each_with_index { |value, index| each_node(value, path + [index], &block) }
  end
end

def sanitize_github_expressions(script)
  script.gsub(/\$\{\{.*?\}\}/m, 'GITHUB_EXPRESSION')
end

def validate_bash_syntax(script, label, errors)
  sanitized = sanitize_github_expressions(script)
  Dir.mktmpdir('workflow-bash-') do |dir|
    path = File.join(dir, 'script.sh')
    File.write(path, "#!/usr/bin/env bash\n#{sanitized}\n")
    _stdout, stderr, status = Open3.capture3('bash', '-n', path)
    return if status.success?

    errors << "#{label}: bash syntax check failed: #{stderr.strip}"
  end
end

workflow_files.each do |file|
  rel = relative(file)
  text = File.read(file)

  if text.match?(/^(<<<<<<<|=======|>>>>>>>)$/)
    errors << "#{rel}: merge conflict marker detected"
  end

  FORBIDDEN_TEXT_PATTERNS.each do |pattern, message|
    next unless text.match?(pattern)

    errors << "#{rel}: #{message}"
  end

  FORBIDDEN_RUNTIME_FLAGS.each do |flag|
    errors << "#{rel}: remove deprecated global runtime flag #{flag}" if text.include?(flag)
  end

  begin
    parsed = YAML.safe_load(text, permitted_classes: [], permitted_symbols: [], aliases: true)
  rescue Psych::SyntaxError => e
    errors << "#{rel}: YAML syntax error at line #{e.line}, column #{e.column}: #{e.problem}"
    next
  end

  each_node(parsed) do |node, path|
    next unless node.is_a?(Hash)

    if node.key?('uses')
      action = node['uses'].to_s
      next if action.start_with?('./', 'docker://')

      unless action.include?('@')
        errors << "#{rel}: #{path.join('.')} uses '#{action}' without a pinned ref"
        next
      end

      ref = action.split('@', 2).last
      if FORBIDDEN_ACTION_REFS.include?(ref)
        errors << "#{rel}: #{path.join('.')} uses unstable action ref '#{action}'"
      end
    end

    next unless node.key?('run')

    shell = node['shell'].to_s
    next if shell.match?(/pwsh|powershell|cmd/i)

    label = "#{rel}: #{path.join('.')} run block"
    validate_bash_syntax(node['run'].to_s, label, errors)
  end
end

Find.find(ROOT) do |path|
  Find.prune if path == File.join(ROOT, '.git') || path == File.join(ROOT, 'node_modules') || path == File.join(ROOT, 'frontend', 'node_modules')
  next unless File.file?(path) && File.extname(path) == '.sh'

  _stdout, stderr, status = Open3.capture3('bash', '-n', path)
  next if status.success?

  errors << "#{relative(path)}: bash syntax check failed: #{stderr.strip}"
end

if errors.any?
  warn errors.map { |error| "::error::#{error}" }.join("\n")
  exit 1
end

puts "OK #{workflow_files.length} GitHub YAML files validated without placeholder/fake-data gates"

File: scripts/validate-github-workflows.js
#!/usr/bin/env node
'use strict';

const fs = require('node:fs');
const path = require('node:path');
const os = require('node:os');
const cp = require('node:child_process');

const ROOT = path.resolve(__dirname, '..');
const GITHUB_DIR = path.join(ROOT, '.github');
const FORBIDDEN_RUNTIME_FLAGS = [
  'FORCE_JAVASCRIPT_ACTIONS_TO_NODE20',
  'FORCE_JAVASCRIPT_ACTIONS_TO_NODE22',
  'FORCE_JAVASCRIPT_ACTIONS_TO_NODE24',
];

const FORBIDDEN_TEXT_PATTERNS = [
  [/ci_dummy|dummy_|fake_|not-configured|your-project\.supabase\.co|your-anon-key|your-public-anon-key|your-publishable-key/i, 'placeholder or dummy credential/data detected'],
  [/continue-on-error:\s*true/i, 'continue-on-error is forbidden for production workflow gates'],
  [/skip_.*=true|skipped.*missing secrets|warning::.*skipped/i, 'skip-on-missing-secrets pattern detected'],
  [/\|\|\s*true/, 'command failure swallowed with || true'],
  [/secrets:\s*inherit/i, 'blanket reusable-workflow secret inheritance is forbidden'],
];

const errors = [];

function relative(filePath) {
  return path.relative(ROOT, filePath).replaceAll(path.sep, '/');
}

function walk(dir, out = []) {
  if (!fs.existsSync(dir)) return out;
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) walk(full, out);
    else if (/\.(ya?ml)$/i.test(entry.name)) out.push(full);
  }
  return out;
}

function commandExists(command) {
  const pathEntries = (process.env.PATH || '').split(path.delimiter).filter(Boolean);
  const extensions = process.platform === 'win32'
    ? (process.env.PATHEXT || '.COM;.EXE;.BAT;.CMD').split(';')
    : [''];

  for (const directory of pathEntries) {
    for (const extension of extensions) {
      const candidate = path.join(directory, `${command}${extension}`);
      try {
        fs.accessSync(candidate, process.platform === 'win32' ? fs.constants.F_OK : fs.constants.X_OK);
        return true;
      } catch {
        // Keep searching PATH.
      }
    }
  }
  return false;
}

function sanitizeGithubExpressions(script) {
  return script.replace(/\$\{\{[\s\S]*?\}\}/g, 'GITHUB_EXPRESSION');
}

function validateBashSyntax(script, label) {
  if (!commandExists('bash')) return;
  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'workflow-bash-'));
  const scriptPath = path.join(tempDir, 'script.sh');
  fs.writeFileSync(scriptPath, `#!/usr/bin/env bash\n${sanitizeGithubExpressions(script)}\n`, 'utf8');
  const result = cp.spawnSync('bash', ['-n', scriptPath], { encoding: 'utf8', shell: false });
  fs.rmSync(tempDir, { recursive: true, force: true });
  if (result.status !== 0) {
    errors.push(`${label}: bash syntax check failed: ${(result.stderr || '').trim()}`);
  }
}

function extractRunBlocks(text) {
  const lines = text.split(/\r?\n/);
  const blocks = [];

  for (let i = 0; i < lines.length; i += 1) {
    const line = lines[i];
    const match = line.match(/^(\s*)run:\s*\|\s*$/);
    if (!match) continue;

    const baseIndent = match[1].length;
    const blockLines = [];
    for (let j = i + 1; j < lines.length; j += 1) {
      const current = lines[j];
      if (current.trim() === '') {
        blockLines.push('');
        continue;
      }
      const indent = current.match(/^\s*/)[0].length;
      if (indent <= baseIndent) break;
      blockLines.push(current.slice(Math.min(indent, baseIndent + 2)));
    }

    blocks.push({ line: i + 1, script: blockLines.join('\n') });
  }

  return blocks;
}

function validateWorkflowFile(file) {
  const rel = relative(file);
  const text = fs.readFileSync(file, 'utf8');
  const lines = text.split(/\r?\n/);

  if (/^(<<<<<<<|=======|>>>>>>>)$/m.test(text)) {
    errors.push(`${rel}: merge conflict marker detected`);
  }

  for (const [pattern, message] of FORBIDDEN_TEXT_PATTERNS) {
    if (pattern.test(text)) errors.push(`${rel}: ${message}`);
  }

  for (const flag of FORBIDDEN_RUNTIME_FLAGS) {
    if (text.includes(flag)) errors.push(`${rel}: remove deprecated global runtime flag ${flag}`);
  }

  lines.forEach((line, index) => {
    if (/^[^#\n]*:\s*[^#\n]+\s{2,}[A-Za-z0-9_-]+:\s*(\||[^#\n]*)?$/.test(line)) {
      errors.push(`${rel}: possible malformed YAML with two keys on line ${index + 1}: ${line.trim()}`);
    }

    const uses = line.match(/^\s*uses:\s*([^\s#]+)\s*$/);
    if (!uses) return;
    const action = uses[1];
    if (action.startsWith('./') || action.startsWith('docker://')) return;
    if (!action.includes('@')) {
      errors.push(`${rel}: line ${index + 1} uses '${action}' without a pinned ref`);
      return;
    }
    const ref = action.split('@').pop();
    if (!/^[0-9a-f]{40}$/i.test(ref)) {
      errors.push(`${rel}: line ${index + 1} external action must be pinned to an immutable 40-char commit SHA: '${action}'`);
    }
  });

  for (const block of extractRunBlocks(text)) {
    validateBashSyntax(block.script, `${rel}: line ${block.line} run block`);
  }
}

const workflowFiles = walk(GITHUB_DIR).sort();
if (workflowFiles.length === 0) {
  errors.push('No GitHub YAML files found under .github/.');
}

for (const file of workflowFiles) validateWorkflowFile(file);

if (errors.length > 0) {
  console.error(errors.map((error) => `::error::${error}`).join('\n'));
  process.exit(1);
}

console.log(`OK ${workflowFiles.length} GitHub YAML files validated with immutable action refs and without placeholder/fake-data gates`);
